### PR TITLE
add multiple reads function with memory parameter

### PR DESCRIPTION
**Pull request by YongpingGao**
add a function called reconstruct_mreads with memory parameter.

To reduce reload times, once the block is loaded, we can get more slices instead of just one slice.

However, the original multiple-reads (get one slice once loaded) can be achieved by setting mem < one big slice bytes(like mem = 0.01G, etc.)

### CODE REVIEW

File: python/imageutils.py [hunk 1]
```diff
@@ -383,40 +384,109 @@ def reconstruct_slice_slice(self, reconstructed, legend, mem=None):
...
+        (Y_size, Z_size, X_size) = rec_dims
+        bytes_per_voxel = self.header['bitpix'] / 8
+        print "The shape of the orignal image is {}".format(rec_dims)
+
+        # extract all the info we needed to reconstruct from the legend file
+        with open(legend, 'r')  as legend:
+            filename_prefix = legend.readline().strip().split('/')[-1].split('_')[0]
+            filename_ext = legend.readline().strip().split(".", 1)[1]
+            blocks_dir = legend.readline().strip().rsplit('/', 1)[0]
+            block_shape = nib.load(legend.readline().strip()).header.get_data_shape()
```
Comment (glatard) [inline python/imageutils.py]: the lines above should probably be put in a separate function, reused in different reconstruction methods.

File: python/imageutils.py [hunk 2]
```diff
@@ -383,40 +384,109 @@ def reconstruct_slice_slice(self, reconstructed, legend, mem=None):
...
+
+            print "filename_prefix: {0}, filename_ext: {1}".format(filename_prefix, filename_ext)
+            print "blocks located in {0}, block's shape is {1}".format(blocks_dir, block_shape)
+
+            legend.seek(0, 0)
+            for i, l in enumerate(legend):
+                pass;
```
Comment (glatard) [inline python/imageutils.py]: fine, but perhaps a bit convoluted?

File: python/imageutils.py [hunk 3]
```diff
@@ -383,40 +384,109 @@ def reconstruct_slice_slice(self, reconstructed, legend, mem=None):
...
+            # n is blocks quatity in each axis. Assume all the blocks in the same size and are all cubic.
```
Comment (glatard) [inline python/imageutils.py]: that doesn't work for BigBrain.

File: python/imageutils.py [hunk 4]
```diff
@@ -383,40 +384,109 @@ def reconstruct_slice_slice(self, reconstructed, legend, mem=None):
...
+            n = (i + 1) ** (1.0 / 3)
```
Comment (glatard) [inline python/imageutils.py]: give more meaningful name or explain what the variables will be used for: `X_size` vs `x_size`, `n`, `n_obs`.

File: python/imageutils.py [hunk 5]
```diff
@@ -383,40 +384,109 @@ def reconstruct_slice_slice(self, reconstructed, legend, mem=None):
...
+
+        (y_size, z_size, x_size) = block_shape
+
+        # before reconstruction. calculate how many slices can be loaded into memory.
+
+        # mem_one_block_one_slice
+        mem_s = (y_size * z_size) * bytes_per_voxel
+        n_obs = n ** 2
+        # number of big slices can fit into the memory: 0 or more
+        n_slices = int((mem / mem_s) // n_obs)
```
Comment (glatard) [inline python/imageutils.py]: not sure why you divide by `n_obs` here.

File: python/imageutils.py [hunk 6]
```diff
@@ -383,40 +384,109 @@ def reconstruct_slice_slice(self, reconstructed, legend, mem=None):
...
+        if n_slices == 0:
+            print "mem is set too low, only one big slice is going to be loaded intto the memory"
+            n_slices = 1
+
+        with open(reconstructed, "wb+") as reconstructed:
+
+            print "writing new header..."
+            with open(self.filepath, 'rb') as f:
+                bytes_arr = f.read(self.header_size)
+                reconstructed.write(bytes_arr)
```
Comment (glatard) [inline python/imageutils.py]: why isn't the header written in other reconstruction methods as well?

File: python/imageutils.py [hunk 7]
```diff
@@ -383,40 +384,109 @@ def reconstruct_slice_slice(self, reconstructed, legend, mem=None):
...
+                            print "{} is loaded into memory".format(block_name)
+                            block_data = nib.load(block_name).get_data(caching='unchanged')
+                            block_data_arr = block_data[:, :, i: (i + n_slices if i + n_slices <= x_size else x_size)]
+                            np_arr_y = block_data_arr if np_arr_y is None else np.vstack((np_arr_y, block_data_arr))
+
+                        np_arr_yz = np_arr_y if np_arr_yz is None else np.hstack((np_arr_yz, np_arr_y))
+
+                    postion = self.header_size + bytes_per_voxel * (
+                        l2 * (n_slices * (Y_size * Z_size)) + l * (x_size * Y_size * Z_size))
+                    reconstructed.seek(postion, 0)
```
Comment (glatard) [inline python/imageutils.py]: you should be able to seek only once every `n_slices` while now you seek much more than that. Also, that part of the code would deserve a little bit more of comments, or better named variables to understand what's going on.

File: python/imageutils.py [hunk 8]
```diff
@@ -383,40 +384,109 @@ def reconstruct_slice_slice(self, reconstructed, legend, mem=None):
...
+                for l2, i in enumerate(range(0, x_size, n_slices)):
+                    np_arr_yz = None
+                    for z in range(0, Z_size, z_size):
+                        np_arr_y = None
+                        for y in range(0, Y_size, y_size):
+                            block_name = blocks_dir + '/' + filename_prefix + "_" + str(y) + "_" + str(z) + "_" + str(
+                                x) + "." + filename_ext
+                            print "{} is loaded into memory".format(block_name)
+                            block_data = nib.load(block_name).get_data(caching='unchanged')
+                            block_data_arr = block_data[:, :, i: (i + n_slices if i + n_slices <= x_size else x_size)]
```
Comment (glatard) [inline python/imageutils.py]: Do we know if nibabel loads have an overhead compared to native ones? In any case, let's keep them consistent all over the code (which I think they are), to allow for meaningful comparisons.

File: python/rec_multi_reads.py [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+import __future__
+import nibabel as nib
+import numpy as np
+import time
+
+# image shape: 3850 3025 3500
+# file name format: bigbrain_3080_1815_1400.nii.gz
+
+
+# info of header. Code should be modified before merging to imageutils.py
```
Comment (glatard) [inline python/rec_multi_reads.py]: yes indeed :)

File: python/rec_multi_reads.py [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+    with open(legend, 'r')  as legend:
+        filename_prefix = legend.readline().strip().split('/')[-1].split('_')[0]
+        filename_ext = legend.readline().strip().split(".", 1)[1]
+        print "filename_prefix: {0}, filename_ext: {1}".format(filename_prefix, filename_ext)
+
+
+    with open(out_image, "ab+") as reconstructed:
+
+        # n blocks each dimension, for now 25 blocks per slice in x dimension
+        n = 5
```
Comment (glatard) [inline python/rec_multi_reads.py]: this can't be hardcoded.

File: python/rec_multi_reads.py [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+# image shape: 3850 3025 3500
+# file name format: bigbrain_3080_1815_1400.nii.gz
+
+
+# info of header. Code should be modified before merging to imageutils.py
+header_size = 352
+bytes_per_voxel = 2
+
+# TODO: Memory, hdfs, other file formats
+def reconstruct_multi_reads(legend, img_shape, blocks_dir, out_image):
```
Comment (glatard) [inline python/rec_multi_reads.py]: I'm assuming this method shouldn't have been pushed here, right?

File: python/rec_multi_reads.py [hunk 12]
```diff
@@ -0,0 +1,79 @@
...
+            reconstructed.seek(header_size + bytes_per_voxel * (l * (x_size // n) * y_size * z_size), 0)
+            print "writing bytes to disk ..."
+            reconstructed.write(np_arr_slice.tobytes('F'))
+            print "{0} level slice is done writing, {1}s is used".format(l, time.time() - one_slice_time)
+
+
+
+def main():
+    start_time = time.time()
+    reconstruct_multi_reads("/data/bigbrain_blocks2/legend.txt", [3850, 3025, 3500], "/data/bigbrain_blocks2", "./recon_mreads.nii")
```
Comment (glatard) [inline python/rec_multi_reads.py]: oops, no hard-coded path, pls.

**Review by glatard — CHANGES_REQUESTED** [review]
Nice work! See my comments in the review.

File: python/imageutils.py (hunk 1 above)
Comment (YongpingGao) [reply python/imageutils.py]: Sure, I will fix it.

File: python/imageutils.py (hunk 2 above)
Comment (YongpingGao) [reply python/imageutils.py]: these lines is just to calculate how many blocks per axis. If not work for Bigbrain, I will figure out another way

File: python/imageutils.py (hunk 3 above)
Comment (YongpingGao) [reply python/imageutils.py]: these lines is just to calculate how many blocks per axis. If not work for Bigbrain, I will figure out another way

File: python/imageutils.py (hunk 4 above)
Comment (YongpingGao) [reply python/imageutils.py]: OK, I will add more comments.

File: python/imageutils.py (hunk 5 above)
Comment (YongpingGao) [reply python/imageutils.py]: I will add more comments on this.

File: python/imageutils.py (hunk 6 above)
Comment (YongpingGao) [reply python/imageutils.py]: I just copy the origin header in bytes and write it to the new image file. I am wondering if there is another way to create the header correctly.

File: python/imageutils.py (hunk 8 above)
Comment (YongpingGao) [reply python/imageutils.py]: Could you be more specific to this one? Do you mean I should use the function "load_image" in imageutils.py to load the image? if it is, I will fix that.

File: python/imageutils.py (hunk 7 above)
Comment (YongpingGao) [reply python/imageutils.py]: Ok I will add more comments.

File: python/rec_multi_reads.py (hunk 9 above)
Comment (YongpingGao) [reply python/rec_multi_reads.py]: rec_multi_reads.py should be removed as it's in the imageutls.py now and I forgot to do that, sry.

File: python/rec_multi_reads.py (hunk 11 above)
Comment (YongpingGao) [reply python/rec_multi_reads.py]: same reason, rec_multi_reads.py should be removed

File: python/rec_multi_reads.py (hunk 10 above)
Comment (YongpingGao) [reply python/rec_multi_reads.py]: same reason, rec_multi_reads.py should be removed

File: python/rec_multi_reads.py (hunk 12 above)
Comment (YongpingGao) [reply python/rec_multi_reads.py]: same reason, rec_multi_reads.py should be removed

**Comment by YongpingGao** [comment]
I have added some replies and I will fix them ASAP

File: python/imageutils.py (hunk 3 above)
Comment (glatard) [reply python/imageutils.py]: I mean, BigBrain blocks are not cubic, will that crash your code?

File: python/imageutils.py (hunk 6 above)
Comment (glatard) [reply python/imageutils.py]: It looks good, but I'm wondering why the header is not written in the other reconstruction  methods in the class. See @ValHayot's comment in Slack.

File: python/imageutils.py (hunk 8 above)
Comment (glatard) [reply python/imageutils.py]: Leave it like this but just note for later that nibabel might introduce an overhead in reading, and that when we compare different methods they should all use nibabel (or none of them should).

**Comment by YongpingGao** [comment]
I have updated the code, let me know if there is more fix needed

File: python/imageutils.py [hunk 13]
```diff
@@ -383,40 +389,104 @@ def reconstruct_slice_slice(self, reconstructed, legend, mem=None):
...
+        # mem_s: how many bytes per slice of one small block
+        mem_s = (y_size * z_size) * bytes_per_voxel
+
+        # n_slices: number of complete slices can fit into the memory: 0 or more
+        # 0 means memory the user set is too low which is smaller than one complete slice (mem_s * ny * nz)
+        # To function properly in mreads, we only allow the minimum of n_slices is 1
+        n_slices = int((mem / (mem_s * ny * nz)))
+        if n_slices == 0:
+            print "mem is set too low, only one complete slice is going to be loaded intto the memory"
+            n_slices = 1
```
Comment (glatard) [inline python/imageutils.py]: Well we should be able to write partial slices if the memory is too low.

File: python/imageutils.py [hunk 14]
```diff
@@ -383,40 +389,104 @@ def reconstruct_slice_slice(self, reconstructed, legend, mem=None):
...
+        (y_size, z_size, x_size) = block_shape
+        # how many blocks on each direction
+        (ny, nz, nx) = (Y_size // y_size, X_size // x_size, Z_size // z_size)
+
+        # before reconstruction. calculate how many slices can be loaded into memory.
+        # mem_s: how many bytes per slice of one small block
+        mem_s = (y_size * z_size) * bytes_per_voxel
+
+        # n_slices: number of complete slices can fit into the memory: 0 or more
+        # 0 means memory the user set is too low which is smaller than one complete slice (mem_s * ny * nz)
```
Comment (glatard) [inline python/imageutils.py]: Great comments, now it is crystal-clear!

File: python/imageutils.py [hunk 15]
```diff
@@ -383,40 +389,104 @@ def reconstruct_slice_slice(self, reconstructed, legend, mem=None):
...
+        # To function properly in mreads, we only allow the minimum of n_slices is 1
+        n_slices = int((mem / (mem_s * ny * nz)))
```
Comment (glatard) [inline python/imageutils.py]: how about `n_slices = mem / (Y_size * Z_size * bytes_per_voxel)?

File: python/imageutils.py [hunk 16]
```diff
@@ -383,40 +389,104 @@ def reconstruct_slice_slice(self, reconstructed, legend, mem=None):
...
+        if n_slices == 0:
+            print "mem is set too low, only one complete slice is going to be loaded intto the memory"
+            n_slices = 1
+
+
+        # l stands for # levels (blocks) on x direction.
+        # x stands for block postion on x direction
+        for l, x in enumerate(range(0, X_size, x_size)):
```
Comment (glatard) [inline python/imageutils.py]: so if the blocks have different x sizes, this will not work.

File: python/imageutils.py [hunk 17]
```diff
@@ -383,40 +389,104 @@ def reconstruct_slice_slice(self, reconstructed, legend, mem=None):
...
+            print("-------------{} level blocks start...-------------".format(l))
+            start_time = time()
+            # l2 stands for how many n_slices in one block (on x direction)
+            for l2, i in enumerate(range(0, x_size, n_slices)):
```
Comment (glatard) [inline python/imageutils.py]: pretty complicated...

File: python/imageutils.py [hunk 18]
```diff
@@ -383,40 +389,104 @@ def reconstruct_slice_slice(self, reconstructed, legend, mem=None):
...
+                        np_arr_y = block_data_arr if np_arr_y is None else np.vstack((np_arr_y, block_data_arr))
+                    # concat np_arr_y on z direction
+                    np_arr_yz = np_arr_y if np_arr_yz is None else np.hstack((np_arr_yz, np_arr_y))
+                # l2 * (n_slices * (Y_size * Z_size)): voxels of part of the blocks (n_slices) have been written to the file.
+                # l * (x_size * Y_size * Z_size): voxels of levels (blocks) have been written to the file.
+                postion = self.header_size + bytes_per_voxel * (l2 * (n_slices * (Y_size * Z_size)) + l * (x_size * Y_size * Z_size))
+                reconstructed.seek(postion, 0)
+                # nifti: Fortran order
+                reconstructed.write(np_arr_yz.tobytes('F'))
+                print("l2: {}, Remeaining memory: {} bytes".format(l2, mem - np_arr_yz.nbytes))
```
Comment (glatard) [inline python/imageutils.py]: I'm not sure if it's correct. I think we can make this all clearer.

**Review by glatard — CHANGES_REQUESTED** [review]
Comments are now perfect, but I think we should revise the logic in mreads.

**Comment by YongpingGao** [comment]
All cluster reads code are belong to Valerie, I copy the code from consider so that I can work based on her code.
This PR is just for your comments, please accept Valerie's PR first, then I will make my PR again.

**Comment by glatard** [comment]
Unsurprisingly, this PR conflicts with master, now that #9 has been merged. @YongpingGao, could you resolve the conflict?

**Comment by YongpingGao** [comment]
I just fix all the conflicts.

File: python/imageutils.py [hunk 19]
```diff
@@ -204,58 +200,127 @@ def clustered_read(self, reconstructed, legend, mem):
...
+            seek_time, defrag_time, write_time = write_to_file(data_dict, reconstructed, bytes_per_voxel)
+
+            total_seek += seek_time
+            total_defrag += defrag_time
+            total_write += write_time
+
+            # clear
+            data_dict = {}
+            # change "next write range" to start next round
+            next_write_range = (next_write_range[0] + mem, next_write_range[1] + mem)
```
Comment (glatard) [inline python/imageutils.py:289]: Shouldn't it be `next_write_range[0]+mem+1`?

**Review by glatard — APPROVED** [review]
Excellent work!

**Pull request merged**